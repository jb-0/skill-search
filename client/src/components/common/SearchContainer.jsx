import React, { useEffect, useState } from 'react';
import Search from './Search';

function SearchContainer() {
  // Store a number of variables in state, firstly the input from the search text box, second the
  // terms that have been added so far and finally the suggested search terms
  const [search, setSearch] = useState({
    searchInputText: '',
    searchTerms: [],
    suggestedTerms: [],
  });

  const permittedTerms = ['node', 'react', 'sql'];

  // Handle changes to the text box and provide suggested terms
  function handleTextBoxUpdates(event) {
    const target = event.target.name;
    const value = event.target.value;

    let suggested;
    if (value.length > 0) { 
      suggested = permittedTerms.filter(word => word.substring(0, value.length) === value);
    } else {
      suggested = [];
    }

    console.log(suggested);
    setSearch((previousValues) => {
      return { ...previousValues, [target]: value, suggestedTerms: suggested };
    });
  }

  // Add items to the search terms array
  function addSearchTerm(event) {
    const term = event.target.id;

    setSearch((previousValues) => {
      return {
        ...previousValues,
        searchTerms: [...previousValues.searchTerms, term],
      };
    });
  }

  // Return the search view and pass the necessary props
  return (
    <Search
      handleTextBoxUpdates={handleTextBoxUpdates}
      addSearchTerm={addSearchTerm}
      suggestedTerms={search.suggestedTerms}
    />
  );
}

export default SearchContainer;
