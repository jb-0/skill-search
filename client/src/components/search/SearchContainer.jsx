import React, { useState } from 'react';
import SearchBox from './SearchBox';
import SearchSuggestion from './SearchSuggestion';
import SearchTerms from './SearchTerms';
import SearchResultsContainer from './SearchResultsContainer';
import DropdownSelect from '../common/DropdownSelect';
import Button from '../common/Button';
import { v4 as uuidv4 } from 'uuid';

// TEMP TERMS DATA TILL MOVED INTO DB
const permittedTerms = [
  'Angular',
  'Babel',
  'CSS',
  'DOM',
  'HTML',
  'Java',
  'JavaScript',
  'Python',
  'React',
  'Node',
  'SQL',
];
const permittedLocations = ['london', 'south east england', 'kent', 'essex'];

function SearchContainer() {
  // Store a number of variables in state, firstly the input from the search text box, second the
  // terms that have been added so far and finally the suggested search terms
  const [search, setSearch] = useState({
    searchInputText: '',
    searchTerms: [],
    suggestedTerms: [],
    submittedSearchTerms: [],
    location: 'kent',
  });

  // Handle changes to the text box and provide suggested terms
  function handleTextBoxUpdates(event) {
    const target = event.target.name;
    const value = event.target.value;

    let suggested;
    if (value.length > 0) {
      suggested = permittedTerms.filter(
        (word) =>
          word.substring(0, value.length).toLowerCase() === value.toLowerCase()
      );
    } else {
      suggested = [];
    }

    setSearch((previousValues) => {
      return { ...previousValues, [target]: value, suggestedTerms: suggested };
    });
  }

  function handleDropDownSelectUpdates(event) {
    const target = event.target.name;
    const value = event.target.value;

    setSearch((previousValues) => {
      return { ...previousValues, [target]: value };
    });
  }

  // Add items to the search terms array
  function addSearchTerm(event) {
    const term = event.target.id;

    // Only add the item if it is not already included
    if (!search.searchTerms.includes(term)) {
      setSearch((previousValues) => {
        return {
          ...previousValues,
          searchTerms: [...previousValues.searchTerms, term],
        };
      });
    }
  }

  // Remove from search terms array
  function removeSearchTerm(event) {
    const term = event.target.id;

    setSearch((previousValues) => {
      const searchTerms = previousValues.searchTerms.filter(
        (word) => word.toLowerCase() !== term.toLowerCase()
      );

      return {
        ...previousValues,
        searchTerms: searchTerms,
      };
    });
  }

  function handleSearchButtonClick() {
    setSearch((previousValues) => {
      return {
        ...previousValues,
        submittedSearchTerms: [...previousValues.searchTerms],
      };
    });
  }

  // Return the search view and pass the necessary props
  return (
    <div>
      <DropdownSelect
        name="location"
        selectOptions={permittedLocations}
        value={search.location}
        onChange={handleDropDownSelectUpdates}
      />

      <SearchBox
        handleTextBoxUpdates={handleTextBoxUpdates}
        suggestedTerms={search.suggestedTerms}
        searchTerms={search.searchTerms}
        searchInputText={search.searchInputText}
      />

      <SearchSuggestion
        suggestedTerms={search.suggestedTerms}
        addSearchTerm={addSearchTerm}
        key={uuidv4()}
      />

      <Button buttonText="Search" buttonAction={handleSearchButtonClick} />

      <SearchTerms
        searchTerms={search.searchTerms}
        removeSearchTerm={removeSearchTerm}
        key={uuidv4()}
      />

      {search.submittedSearchTerms.length > 0 ? (
        <SearchResultsContainer searchTerms={search.submittedSearchTerms} location={search.location} />
      ) : null}
    </div>
  );
}

export default SearchContainer;
