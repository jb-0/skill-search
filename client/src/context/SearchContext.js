/**
  * This context is used to pass state to the numerous components and events (such as save) that
  * make up a search. This also means a user can leave the search screen to view their profile
  * without losing search information.
  */
import React, { createContext, useState  } from 'react';

export const SearchContext = createContext();

export const SearchProvider = (props) => {
  // Store a number of variables in state, firstly the input from the search text box, second the
  // terms that have been added so far and finally the suggested search terms
  const [search, setSearch] = useState({
    searchInputText: '',
    searchTerms: [],
    suggestedTerms: [],
    submittedSearchTerms: [],
    location: 'london',
    submittedLocation: '',
    handleDropDownSelectUpdates,
    handleSearchButtonClick
  });

  function handleDropDownSelectUpdates(event) {
    const target = event.target.name;
    const value = event.target.value;

    setSearch((previousValues) => {
      return { ...previousValues, [target]: value };
    });
  }

  // When a user hits search ensure the location and searchTerms are current
  function handleSearchButtonClick() {
    setSearch((previousValues) => {
      return {
        ...previousValues,
        submittedSearchTerms: [...previousValues.searchTerms],
        submittedLocation: previousValues.location,
      };
    });
  }
  
  return <SearchContext.Provider value={ [search, setSearch] }>{props.children}</SearchContext.Provider>;
};
