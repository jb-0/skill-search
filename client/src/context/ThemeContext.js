/**
 * The theme context provides styling theme attributes as well as defining global styles
 */
import React, { createContext } from 'react';
import { createGlobalStyle } from 'styled-components';

export const ThemeContext = createContext();

export const ThemeProvider = (props) => {
  const theme = {
    primaryColor: '#212529',
    secondaryColor: 'white',
    tertiaryColor: '#ffb703',
    primaryPad: '20px',
    mainFont: `Nunito', sans-serif`,
  };

  return (
    <ThemeContext.Provider value={theme}>
      {props.children}
    </ThemeContext.Provider>
  );
};

export const GlobalStyle = createGlobalStyle`
  body {
  margin: 0;
  font-family: var(--main-font);
  text-align: center;
  color: var(--primary-color);
  background-color: var(--primary-color);
  margin-bottom: 60px;
  }

  p,
  input,
  select,
  textarea,
  button {
    font-family: var(--main-font);
    color: var(--primary-color);
    font-size: 16px;
  }

  h1 {
    font-size: 2.5rem;
    font-weight: 900;
  }

  a {
    text-decoration: none;
  }

  .large-p, .large-a {
    font-size: 22px;
  }

  .secondary-font-color {
    color: var(--secondary-color);
  }
`;
