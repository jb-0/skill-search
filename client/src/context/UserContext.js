import React, { createContext, useEffect, useState } from 'react';

export const UserContext = createContext({
  isLoggedIn: undefined,
});

export const UserProvider = (props) => {
  const [userData, setUserData] = useState({
    isLoggedIn: undefined,
  });

  async function checkLoggedIn() {
    const res = await fetch(
      '/api/user/isloggedin',
      {
        method: 'GET',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          frontend: 'react-frontend',
          credentials: 'include'
        },
      }
    );
    
    // Fail safe, if for any reason we cannot reach the server or the request is bad and a non 200
    // is returned we switch to false to ensure private routes remain private.
    if (res.status === 200) {
      setUserData({ isLoggedIn: await res.json() });
    } else {
      setUserData({ isLoggedIn: false });
    }
  }

  useEffect(() => {
    checkLoggedIn();
  }, []);

  return <UserContext.Provider value={ userData }>{props.children}</UserContext.Provider>;
};


