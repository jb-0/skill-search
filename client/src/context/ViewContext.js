import React, { createContext, useState, useLayoutEffect } from 'react';

export const ViewContext = createContext({
  width: window.innerWidth,
  height: window.innerHeight,
  device: window.innerWidth <= 480 ? 'Mobile' : 'Desktop'
});

export const ViewProvider = (props) => {
  const [size, setSize] = useState({
    width: window.innerWidth,
    height: window.innerHeight,
    device: window.innerWidth <= 480 ? 'Mobile' : 'Desktop'
  });

  useLayoutEffect(() => {
    function updateSize() {
      setSize({ 
        width: window.innerWidth, 
        height: window.innerHeight, 
        device: window.innerWidth <= 480 ? 'Mobile' : 'Desktop' 
      });
    }

    window.addEventListener('resize', updateSize);
    updateSize();

    return () => window.removeEventListener('resize', updateSize);
  }, []);
  return <ViewContext.Provider value={ size }>{props.children}</ViewContext.Provider>;
};
