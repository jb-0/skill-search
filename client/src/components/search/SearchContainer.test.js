import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

import SearchContainer from './SearchContainer';
import { SearchProvider } from '../../context/SearchContext';

describe('SearchContainer component', () => {
  // let container = null;
  // beforeEach(() => {
  //   // setup a DOM element as a render target
  //   container = document.createElement('div');
  //   document.body.appendChild(container);

  //   act(() => {
  //     render(<SearchProvider><SearchContainer /></SearchProvider>, container);
  //   });
  // });

  // afterEach(() => {
  //   // cleanup on exiting
  //   unmountComponentAtNode(container);
  //   container.remove();
  //   container = null;
  // });

  it('combined search components render correctly with default state', () => {
    const tree = renderer
      .create(<SearchProvider><SearchContainer /></SearchProvider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
