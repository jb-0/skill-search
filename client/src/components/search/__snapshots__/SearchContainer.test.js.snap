// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchContainer component combined search components render correctly with default state 1`] = `
<div>
  <div
    className="dropdown-select"
  >
    <select
      id="location"
      name="location"
      onChange={[Function]}
      value="london"
    >
      <option
        value="london"
      >
        london
      </option>
      <option
        value="south east england"
      >
        south east england
      </option>
      <option
        value="kent"
      >
        kent
      </option>
      <option
        value="essex"
      >
        essex
      </option>
    </select>
  </div>
  <div
    className="search-container"
  >
    <input
      autoComplete="off"
      id="searchInputText"
      name="searchInputText"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div />
  <button
    className="base-button undefined"
    onClick={[Function]}
    type="button"
  >
    Search
  </button>
  <div />
</div>
`;
